src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -3650,14 +3654,31 @@ def _array_syntax_to_indexed(self, parent, loop_vars):
             else:
                 first_rank = rank
 
-            # Replace the PSyIR Ranges with the loop variables
+            base_ref = _copy_full_base_reference(array)
+            array_ref = array.ancestor(Reference, include_self=True)
+            shape = array_ref.datatype.shape
+            add_op = BinaryOperation.Operator.ADD
+            sub_op = BinaryOperation.Operator.SUB
+            # Replace the PSyIR Ranges with appropriate index expressions.
             range_idx = 0
             for idx, child in enumerate(array.indices):
                 if isinstance(child, Range):
-                    symbol = _find_or_create_unresolved_symbol(
-                        array, loop_vars[range_idx],
-                        symbol_type=DataSymbol, datatype=DeferredType())
-                    array.children[idx] = Reference(symbol)
+                    symbol = table.lookup(loop_vars[range_idx])
+                    if isinstance(shape[range_idx], ArrayType.Extent):
+                        # We don't know the bounds of this array so we have
+                        # to query using LBOUND.
+                        lbound = IntrinsicCall.create(
+                            IntrinsicCall.Intrinsic.LBOUND,
+                            [base_ref.copy(),
+                             ("dim", Literal(str(idx+1), INTEGER_TYPE))])
+                    else:
+                        lbound = shape[range_idx].lower.copy()
+                    # Create the index expression:
+                    #    idx-expr = array-lower-bound + loop-idx - 1
+                    expr = BinaryOperation.create(
+                        add_op, lbound, Reference(symbol))
+                    expr2 = BinaryOperation.create(sub_op, expr, one.copy())
+                    array.children[idx] = expr2
```

Comment:
Although this is technically correct, most of the time array-lower-bound is 1 and we end up with "1 + loop-idx - 1". Could we create a special case for this to avoid the "+1-1" clutter

Reply:
Done.